[FEATURE] Activate an extension including its dependencies

oemmes opened the pull request:
In my composer project I have a distribution extension which requires and depends on other extensions. When I install that extension via the Extension Manager in the backend all dependent extensions are installed, too.

I thought this would be a nice feature for typo3_console as well.

Usage: typo3cms extension:activate --with-dependencies my_extension

The code used for it is taken (with some adjustments) from [\TYPO3\CMS\Extensionmanager\Controller\ActionController](https://github.com/TYPO3/TYPO3.CMS/blob/master/typo3/sysext/extensionmanager/Classes/Controller/ActionController.php#L80)

mbrodala commented on Classes/Command/ExtensionCommandController.php:
```diff
@@ -34,11 +34,21 @@ class ExtensionCommandController extends CommandController
     /**
+     * @var \TYPO3\CMS\Extensionmanager\Service\ExtensionManagementService
+     */
+    protected $managementService;
```
Please rename to `extensionManagementService`

helhum commented:
Change looks good so far. Can you please rebase against master and squash all changes in to one commit?

oemmes commented:
I've rebased the change and squashed the commits. (Had to recommit twice because I mistakenly committed with a wrong mail adress...)

I hoped to get this in for the 2.0.0 release to use it in our projects. As you just released version 2.0.0, when can I expect this to be in a release?

btw: thanks for the extension! it helps us a lot :)

mbrodala commented:
@helhum I think the squashing can be done by GH: https://help.github.com/articles/configuring-pull-request-merge-squashing/

helhum commented:
> I hoped to get this in for the 2.0.0 release to use it in our projects. As you just released version 2.0.0, when can I expect this to be in a release?

2.0.0 is just a maintenance release on top of 1.3.0 (see https://github.com/helhum/typo3_console/releases/tag/2.0.0) to get closer to semantic versioning.

3.0.0 will follow soon

helhum commented:
> I hoped to get this in for the 2.0.0 release to use it in our projects

Can you outline your exact use case? I would be interested, as I never needed this feature and will never need it in my workflow.

oemmes commented:
> an you outline your exact use case?

Sure.

We have a product which uses Typo3 as backend which we sell to our customers. When we set up a new customer project on our server we use composer to "install" a Typo3 site. After that we use typo3_console:install to configure it.

We always need the same page tree, some files etc. which we added into a distribution extension. That distribution extension has also dependencies to other required extensions defined which must be activated in Typo3 for our product to work.

Without this new feature for typo3_console we had to have a list of all that extensions to activate them or we had to activate the distribution in the Backend/ExtensionManager (which activates the dependencies).

The new feature allows us to finish the set up of a project just by activating the distribution extension from the console and don't have to care which dependencies are really in there.

helhum commented:
> The new feature allows us to finish the set up of a project just by activating the distribution extension from the console and don't have to care which dependencies are really in there.

The case you describe is exactly what I also need and what I've built some tools into typo3_console. With these tools, in place the feature in this pull request is obsolete. Here is what I am doing:

**tl;dr:** Define which TYPO3 system extensions should be active (as all come bundled, but you never need every single one) and make all additionally (composer) installed extensions active. Then setup/ configure all active extensions.

Here is the how:

1. Install TYPO3 from command line using typo3_console
1. Do not commit typo3conf/PackageStates.php (commit other files at will)
1. Define the environment variable `TYPO3_ACTIVE_FRAMEWORK_EXTENSIONS` with a comma separated list of system extensions to be active
1. Call `./typo3cms install:generatepackagestates` (both steps can be [combined](https://github.com/helhum/TYPO3-Distribution/blob/master/composer.json#L29-L31) with composer scripts and then running `composer run package-states`)
1. Call `./typo3cms extension:setupactive`

This can be used as template for **any** project. No need to hard code and (site) extension to be setup with its dependencies.

What do you think? You still need this feature here?

helhum commented:
@oemmes don't get me wrong, we may still get this feature in. However especially in composer mode it might be weird, because the code might trigger a TER download of an extension, which might break something. I'm just interested if you could live without it in 3.0.0 as I think this needs some polishing still and I want to get 3.0.0 out of the door

oemmes commented:
@helhum sorry for not answering, I had some other tasks to do yesterday. I will give it a try today using your workflow

oemmes commented:
So... your workflow is also working. Though I see the framework extension list as a little overhead regarding our projects.

> in composer mode it might be weird, because the code might trigger a TER download of an extension

I thought about that and you're right. There shouldn't be any (automatic) download of extensions.
Probably the feature shouldn't use the code I just copied from the ExtensionManager, that was just the easiest way ;) It should only try to activate the dependencies - if one of them is not available locally it will fail anyway.
Maybe a version check could be performed beforehand - but actually that should have been taken care of by composer install/update already.

What about non-composer mode? Is typo3_console compatible with that? Then the download of dependend extensions might be valid.

So for now we'll stick with your workflow and I might update the "--with-dependencies" feature to work as described above. Or do you have further suggestions?

helhum commented:
> Though I see the framework extension list as a little overhead regarding our projects

How come? Do you currently go with what TYPO3 activates by default during installations?

helhum commented:
> What about non-composer mode? Is typo3_console compatible with that? Then the download of dependend extensions might be valid.

true. So one additional thing to consider here.

> So for now we'll stick with your workflow and I might update the "--with-dependencies" feature to work as described above.

Cool, thanks!

>  Or do you have further suggestions?

No. I only do not want to integrate something half baked, that I don't need but have to maintain lateron.

To take a final decision here, I'll close this PR and suggest that you integrate this feature in a minimal extension first (which works nicely with the console as you can now even register lowlevel commands for extensions). (Of course it would be a new command and not an additional argument then.)

Would be great if you could publish this ext on github. If you do so, you could use this feature in your projects by additionally requiring the extension, without any further changes that need to be made here.

oemmes commented:
> > Though I see the framework extension list as a little overhead regarding our projects

> How come? Do you currently go with what TYPO3 activates by default during installations?

Yes, that worked for us as our users don't have to use the backend extensively.

> I'll close this PR and suggest that you integrate this feature in a minimal extension first
> Would be great if you could publish this ext on github.

That's okay for me. Though I have to see if we'll stick with the list of framework extensions or eventually need to have an "activate an extension with dependencies" command.

helhum commented:
> That's okay for me. 

Thanks!

> Though I have to see if we'll stick with the list of framework extensions or eventually need to have an "activate an extension with dependencies" command.

Let me know how you decide.

helhum commented:
@oemmes 

> Though I see the framework extension list as a little overhead regarding our projects

With the `--activate-default-extensions` switch, you can now safely omit the active framework extension configuration and get those activated, which would be active after a regular TYPO3 setup.

See: https://github.com/helhum/typo3_console/pull/162
